mailcom/parse.py

```diff
@@ -10,86 +10,247 @@
+            entity["replaced"] = True
+            ent_string = entity["entity"]  # noqa
+            # here we could check that string is "I-PER"
+            ent_conf = entity["score"]  # noqa
+            ent_position = entity["start"], entity["end"]
+            # Here we have to be careful - tokenization with
+            # transformers is quite different from spacy/stanza/flair
+            # here we get character ids
+            entlist.append(ent_position)
+            # now replace respective characters
+            # replace I-PER
+            if ent_string == "I-PER":
+                # add all following entities to this name
+                word_end = entity["end"]
+                for j in range(i, len(ner) - 1):
+                    if ner[j]["end"] != ner[j + 1]["start"]:
+                        break
+                    ner[j + 1]["replaced"] = True
+                    word_end = ner[j + 1]["end"]
+                name_to_replace = new_sentence[
+                    entity["start"]
+                    + additional_sentence_length : word_end  # noqa
+                    + additional_sentence_length
+                ]
+                # distinguish between first and last names
+                is_last_name = False
+                # if there is another entity before this name with one char inbetween,
+                # this is likely a last name
+                if i > 0 and entity["start"] - ner[i - 1]["end"] == 1:
+                    is_last_name = True
+                # if there is a form of address in front
+                # of this name that indicates last names,
+                # this is likely a last name
+                region_to_search_for_foa = new_sentence[
+                    entity["start"] - 10 : entity["start"]  # noqa
+                ]
+                if any(
+                    foa in region_to_search_for_foa
+                    for foa in self.ln_forms_of_adress
+                ):
+                    is_last_name = True
+                # choose the pseudonym
+                if is_last_name:
+                    nm_list = self.used_last_names
+                    pseudo_list = self.pseudo_last_names
+                else:
+                    nm_list = self.used_first_names
+                    pseudo_list = self.pseudo_first_names
+                pseudonym = ""
+                name_variations = [
+                    name_to_replace,
+                    name_to_replace.lower(),
+                    name_to_replace.title(),
+                ]
+                # if this name has been replaced before, choose the same pseudonym
+                for nm_var in name_variations:
+                    pseudonym = nm_list.get(nm_var, "")
+                    if pseudonym != "":
+                        break
+                # if none is found, choose a new pseudonym
+                if pseudonym == "":
+                    pseudonym = pseudo_list["fr"][len(nm_list)]  # reaches end of the list
```

Comment:
This potentially reaches the end of the list and breaks. Hopefully this disappears when you refactor the complexity.